src/main/scala/chisel3/simulator/PeekPokeAPI.scala

```diff
@@ -127,58 +304,236 @@ trait PeekPokeAPI {
+      )
+      val peekedValue = data.peek()
+      data.zip(expected).zipWithIndex.foreach { case ((datEl, expEl), index) =>
+        expEl match {
+          case DontCare =>
+          // TODO: missing elements?
+          case exp if exp.getClass == datEl.getClass =>
+            datEl.expect(
+              exp.asInstanceOf[T],
+              (obs: T, exp: T) => buildMessage(peekedValue, expected, index)
+            )
+          case _ =>
+            throw new Exception(
+              s"Type mismatch: ${expEl.getClass} != ${datEl.getClass}"
+            )
         }
       }
     }
+
+    override def expect(expected: Vec[T], buildMessage: (Vec[T], Vec[T]) => String)(
+      implicit sourceInfo: SourceInfo
+    ): Unit =
+      expectVec[T](
+        expected,
+        (observed: Vec[T], expected: Vec[T], idx: Int) =>
+          buildMessage(observed, expected) + s". First mismatch at index $idx."
+      )
+  }
+
+  implicit class TestableData[T <: Data](val data: T) extends PeekPokable[T] {
+
+    def peek(): T = {
+      data match {
+        case x: Bool     => new TestableBool(x).peek().asInstanceOf[T]
+        case x: UInt     => new TestableUInt(x).peek().asInstanceOf[T]
+        case x: SInt     => new TestableSInt(x).peek().asInstanceOf[T]
+        case x: EnumType => new TestableEnum(x).peek().asInstanceOf[T]
+        case x: Record   => new TestableRecord(x).peek().asInstanceOf[T]
+        case x: Vec[_]   => new TestableVec(x).peek().asInstanceOf[T]
+        case x => throw new Exception(s"don't know how to peek $x")
+      }
+    }
+
+    override def expect(
+      expected:     T,
+      buildMessage: (T, T) => String
+    )(implicit sourceInfo: SourceInfo): Unit = {
+
+      def buildMsgFn[S](obs: S, exp: S): String = {
+        require(obs.getClass == exp.getClass, s"Type mismatch: ${obs.getClass} != ${exp.getClass}")
+        buildMessage(obs.asInstanceOf[T], exp.asInstanceOf[T])
+      }
+
+      (data, expected) match {
+        case (dat: Bool, exp: Bool) =>
+          new TestableBool(dat).expect(exp, buildMsgFn _)
+        case (dat: UInt, exp: UInt) =>
+          new TestableUInt(dat).expect(exp, buildMsgFn _)
+        case (dat: SInt, exp: SInt) =>
+          new TestableSInt(dat).expect(exp, buildMsgFn _)
+        case (dat: EnumType, exp: EnumType) =>
+          new TestableEnum(dat).expect(exp, buildMsgFn _)
+        case (dat: Record, exp: Record) =>
+          new TestableRecord(dat).expect(exp, buildMsgFn _)
+        case (dat: Vec[_], exp: Vec[_]) =>
+          new TestableVec(dat).expectVec(
+            exp,
+            (obs: Data, _: Data, idx: Int) => {
+              require(obs.getClass == exp.getClass, s"Type mismatch: ${obs.getClass} != ${exp.getClass}")
+              buildMessage(obs.asInstanceOf[T], expected.asInstanceOf[T]) + s". First mismatch at index $idx."
+            }
+          )
```

Comment:
I'm a little confused why there is special logic for Vec here but not for Record? I see Records are checked in the methods on `TestableRecord` so I think it's fine there's no extra checks, but Vecs are tested in `TestableVec` so why are there additional checks here?

Reply:
It's probably not needed, but I wasn't 100% sure. The type generics for the extension classes differ between `Vec` and `Record`:
```scala
class TestableRecord[T <: Record](val data: T)  // type parameter is the type itself, so methods accepting T are type-checked
```

vs
```scala
class TestableVec[T <: Data](val data: Vec[T]) extends PeekPokable[Vec[T]]  // type parameter is the element type and is gone after type erasure
```
That's the only way I could get `Vec[T]` working. Please let me know if you think there's a cleaner way.

I cleaned up the code a little bit (using a weird workaround for `Vec` in `_expect` and `_poke`), so the special cases are now tucked away.